### ISSUE TITLE
Rounded Progressbar does not behave well for values below the corner radius

### ISSUE THREAD

**Issue by IceflowRE** [godotengine/godot#87690]
### Tested versions

v4.2.1.stable.official [b09f793f5]

### System information

Godot v4.2.1.stable - Windows 10.0.22631 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1080 Ti (NVIDIA; 31.0.15.4633) - AMD Ryzen 9 7900X 12-Core Processor (24 Threads)

### Issue description

The progressbar with a StyleBoxFlat is distorted for values below the corner radius.
![image](https://github.com/godotengine/godot/assets/10883690/c00bf392-55fd-4bb3-9515-5ff5c8fe824b)
For larger values it is ok again
![image](https://github.com/godotengine/godot/assets/10883690/7ba3fd4b-7f56-4a95-8395-be7e5b46fc3e)

The stylebox needs to be cut off at that value, instead of being distorted.

Related to https://github.com/godotengine/godot/issues/44015, but with StyleBoxTexture.

### Steps to reproduce

Create Progressbar with a StyleBox who has larger rounded corners, to make it visible.

### Minimal reproduction project (MRP)

N/A

**Comment by Calinou** [comment]
> The stylebox needs to be cut off at that value, instead of being distorted.

Unfortunately, the stylebox can't know which way it should be cut in. It could be either the left or right side - there's no way for the stylebox to do this correctly without knowing the *context* in which it's being placed.

**Comment by duongxinh2003** [comment]
If I use TextureProgressBar following  @Calinou , at the end of progress was cut and not bordered like this 
<img width="318" alt="image" src="https://github.com/user-attachments/assets/511a0971-35fe-4f3b-9c07-9070d78114c1">
Do you have any way to round the tail?

**Comment by Calinou** [comment]
>  Do you have any way to round the tail?

No, TextureProgressBar only supports "straight" clipping. If you want a rounded end, use ProgressBar with a StyleBoxTexture for the background and foreground. You'll need to configure the StyleBoxTexture to use ninepatch-style stretching.

**Comment by JRDK92** [comment]
> > The stylebox needs to be cut off at that value, instead of being distorted.
> 
> Unfortunately, the stylebox can't know which way it should be cut in when it's more narrow than its corner radius. It could be _any_ of the sides, including top/bottom - there's no way for the stylebox to do this correctly without knowing the _context_ in which it's being placed. We can't assume the progress bar is always being filled from left to right, as Godot supports all directions (including bilinear progress bars that expand from the middle).
> 
> Also, people may be relying on the existing behavior by having very high corner radii set to create "pill" buttons regardless of their size. This is like `border-radius: 9999px;` in CSS.
> 
> The only solution I can see is to add four boolean StyleBoxFlat properties which allows you to control which edges should be "cut" in this situation. This is a lot of properties for an use case that isn't very common (compared to general StyleBoxFlat usage) 🙁
> 
> As a workaround, use a TextureProgressBar node for now.

What's the solution for this 2 years later? Godot 4.6 stable, still seems to happen and I can't figure out how to fix it.

**Comment by duongxinh2003** [comment]
Is there any update on this? I've been waiting for more than two years for this bug to be fixed.

**Comment by Calinou** [comment]
> Could the background clip the fill element?

This could work using CanvasItem clipping, but it's quite expensive and not an optimal solution. Doing this unconditionally on every ProgressBar node would add a significant cost to rendering.

Clipping could be made opt-in via a property, but I still feel a better solution is to modify StyleBoxFlat to add properties to control how rounded corners behave when the StyleBox is too small to fit them.
